fenapack/field_split_backend.py

```diff
@@ -117,24 +121,24 @@ def get_work_dolfin_mat(self, key, comm,
         return dolfin_mat
 
 
-    def setup_ksp_once(self, ksp, assemble_func, iset, spd=False):
+    def setup_ksp(self, ksp, assemble_func, iset, spd=False, const=False):
         """Assemble into operator of given ksp if not yet assembled"""
         mat = ksp.getOperators()[0]
-        # FIXME: This logic that it is created once should be visible
-        #        in higher level, not in these internals
+        prefix = ksp.getOptionsPrefix()
         if mat.type is None or not mat.isAssembled():
             # Assemble matrix
+            destruction = True if const else None
             dolfin_mat = self.get_work_dolfin_mat(assemble_func, mat.comm,
-                                                  can_be_destroyed=True,
+                                                  can_be_destroyed=destruction,
                                                   can_be_shared=True)
             assemble_func(dolfin_mat)
             mat = self._get_deep_submat(dolfin_mat, iset, submat=None)
+            # NOTE: shallow version results in PETSc error 92 [MatGetFactor() failed]
```

Comment:
Why should this work? Cholesky needs assembled matrix, AMG needs assembled matrix, ... We expect shallow submats to work only for operators used for multiplications rather than solves. If course it could work for some real matrix-free method.

Reply:
Sure, got it...